src/planner/planner.c

```diff
@@ -678,35 +675,47 @@ get_or_add_baserel_from_cache(Oid chunk_reloid, Oid parent_reloid)
 		return entry;
 	}
 
-	/*
-	 * This reloid is not in the chunk cache, so do the full metadata
-	 * lookup.
-	 */
-	int32 hypertable_id = 0;
-	int32 chunk_status = 0;
-	if (ts_chunk_get_hypertable_id_and_status_by_relid(chunk_reloid, &hypertable_id, &chunk_status))
+	if (OidIsValid(parent_reloid))
 	{
-		/*
-		 * This is a chunk. Look up the hypertable for it.
-		 */
-		if (OidIsValid(parent_reloid))
+		ht = ts_planner_get_hypertable(parent_reloid, CACHE_FLAG_CHECK);
+
+#ifdef USE_ASSERT_CHECKING
+		/* Sanity check on the caller-specified hypertable reloid. */
+		int32 parent_hypertable_id = ts_chunk_get_hypertable_id_by_relid(chunk_reloid);
+		if (OidIsValid(parent_hypertable_id))
 		{
-			/* Sanity check on the caller-specified hypertable reloid. */
-			Assert(ts_hypertable_id_to_relid(hypertable_id) == parent_reloid);
+			Assert(ts_hypertable_id_to_relid(parent_hypertable_id) == parent_reloid);
+
+			if (ht != NULL)
+			{
+				Assert(ht->fd.id == parent_hypertable_id);
+			}
 		}
-		else
+#endif
+	}
+	else
+	{
+		/* Hypertable reloid not specified by the caller, look it up by
+		 * an expensive metadata scan.
+		 */
+		int32 hypertable_id = ts_chunk_get_hypertable_id_by_relid(chunk_reloid);
+
+		if (OidIsValid(hypertable_id))
 		{
 			/* Hypertable reloid not specified by the caller, look it up. */
 			parent_reloid = ts_hypertable_id_to_relid(hypertable_id);
+
+			if (OidIsValid(parent_reloid))
```

Comment:
Assert(parent_reloid)? If we have a valid hypertable id, we should have a valid reloid?

Reply:
Good point, changed to an `Assert`.

Reply:
> Assert(parent_reloid)? If we have a valid hypertable id, we should have a valid reloid?

... only if the metadata is not corrupt, which could happen for a variety of reasons. Might be good to use `Ensure` here to prevent the release build from crashing on corrupt metadata. I note that the `parent_reloid` is used to look up the hypertable, which is then cached without checking if it is null. This could lead to hard-to-find errors and lead to crashes in debug builds.

Reply:
That's a good hint. I will change the `Assert` to an `Ensure` in a follow-up PR.